File: 0739-daily-temperatures/0739-daily-temperatures.cpp
// same as find next greater number 
// using stack time O(n) space O(n)
class Solution {
public: 
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        stack<pair<int,int>> s; // value - index 
        for(int i = 0; i<temperatures.size(); i++){
            int curr_number = temperatures[i];
            while(!s.empty() && curr_number > s.top().first){
                temperatures[s.top().second] = i-s.top().second;
                s.pop();
            }
            s.push({curr_number,i});
        }
        while(!s.empty()){
            temperatures[s.top().second] = 0;
            s.pop();
        }
    return temperatures;
    }
};

File: 0739-daily-temperatures/0739-daily-temperatures.js
/**
 * @param {number[]} temperatures
 * @return {number[]}
 */
var dailyTemperatures = function(temperatures) {
    const stack = new Array(); // [ [value,index] ]
    temperatures.forEach(function(value,index,array) {
        // current number = value
        while(stack.length && value > stack.at(-1)[0]){
            array[stack.at(-1)[1]] = index - stack.pop()[1]; 
        }
        stack.push([value,index]);
    })
    while(stack.length) temperatures[stack.pop().at(-1)] = 0;
    return temperatures;
};